### ISSUE TITLE
Support other means of Webpack sockets

### ISSUE THREAD

**Issue by sibelius** [pmmmwh/react-refresh-webpack-plugin#10]
```jsx
Uncaught TypeError: connection.onClose is not a function
    at createSocket (webpack-internal:///../../node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/createSocket.js:32)
    at eval (webpack-internal:///../../node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js:80)
    at Object.../../node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js (bundle.js:4276)
    at __webpack_require__ (bundle.js:736)
```

this is the problem

```jsx
let SocketClient;
if (typeof __webpack_dev_server_client__ !== 'undefined') {
  SocketClient = __webpack_dev_server_client__;
} else {
  SocketClient = __webpack_require__("../../node_modules/sockjs-client/lib/entry.js");
}
```

we are using https://github.com/shellscape/webpack-plugin-serve instead of https://github.com/webpack/webpack-dev-server

maybe @shellscape or @PlayMa256 can help us with this one

**Comment by shellscape** [comment]
@sibelius this module looks to be specifically tailored to webpack-dev-server. WPS doesn't use sockjs and we don't export WDS variables, so this module will probably have to allow for detection of WPS and use the socket provided.

**Comment by pmmmwh** [comment]
> ```js-jsx
> Uncaught TypeError: connection.onClose is not a function
>     at createSocket (webpack-internal:///../../node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/createSocket.js:32)
>     at eval (webpack-internal:///../../node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js:80)
>     at Object.../../node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js (bundle.js:4276)
>     at __webpack_require__ (bundle.js:736)
> ```
> 
> this is the problem
> 
> ```js-jsx
> let SocketClient;
> if (typeof __webpack_dev_server_client__ !== 'undefined') {
>   SocketClient = __webpack_dev_server_client__;
> } else {
>   SocketClient = __webpack_require__("../../node_modules/sockjs-client/lib/entry.js");
> }
> ```
> 
> we are using [shellscape/webpack-plugin-serve](https://github.com/shellscape/webpack-plugin-serve) instead of [webpack/webpack-dev-server](https://github.com/webpack/webpack-dev-server)
> 
> maybe @shellscape or @PlayMa256 can help us with this one

Yes - for the current implementation, I purposefully stick to what CRA is doing so as to allow easier integration with them. I will need to look at the socket implementation for WPS and come up with a way to allow users to customize this function. Thanks for raising this though - I was planning to support custom sockets anyways.

**Comment by tim-soft** [comment]
I'm also seeing this with wepback-dev-middleware + webpack-hot-middleware + express

**Comment by pmmmwh** [comment]
Support for Webpack Hot Middleware landed #23 !
